### DISCUSSION TITLE
Airflow providers installation for connections

### DISCUSSION THREAD

**Discussion by Kelvinchow79** [apache/airflow#30381] [Q&A] [answered]
I have set up a airflow 2.5.1, using docker compose in an EC2. I have already migrated all connections from my previous airflow 1.10, and installed most of the packages. But i have some problems with some connections. I have installed each in scheduler, worker and webserver.

For HDFS connection, I have installed; pip install snakebite-py3 pip install apache-airflow-providers-apache-hdfs but installation ends in error, as pip only state package sub-process problem

For Snowflake connection, I have installed; pip install apache-airflow-providers-snowflake pip install apache-airflow-providers-snowflake[common.sql] But it still wont appear in the connection type list

For Hive Client Wrapper, Hive Server 2 Thrift, Hive Metastore Thrift pip install apache-airflow-providers-apache-hive pip install apache-airflow-providers-apache-hive[amazon] pip install thrift_hive_metastore but all installation ends in error, as pip only state package sub-process problem

Error:

Requirement already satisfied: pure-sasl>=0.6.2 in /home/airflow/.local/lib/python3.7/site-packages (from thrift-sasl>=0.1.0->pyhive[hive]>=0.6.0->apache-airflow-providers-apache-hive->apache-airflow[hive]) (0.6.2) Building wheels for collected packages: sasl Building wheel for sasl (setup.py) ... error error: subprocess-exited-with-error

× python setup.py bdist_wheel did not run successfully. │ exit code: 1 ╰─> [26 lines of output] running bdist_wheel running build running build_py creating build creating build/lib.linux-x86_64-3.7 creating build/lib.linux-x86_64-3.7/sasl copying sasl/init.py -> build/lib.linux-x86_64-3.7/sasl running egg_info writing sasl.egg-info/PKG-INFO writing dependency_links to sasl.egg-info/dependency_links.txt writing requirements to sasl.egg-info/requires.txt writing top-level names to sasl.egg-info/top_level.txt reading manifest file 'sasl.egg-info/SOURCES.txt' reading manifest template 'MANIFEST.in' adding license file 'LICENSE.txt' writing manifest file 'sasl.egg-info/SOURCES.txt' copying sasl/saslwrapper.cpp -> build/lib.linux-x86_64-3.7/sasl copying sasl/saslwrapper.h -> build/lib.linux-x86_64-3.7/sasl copying sasl/saslwrapper.pyx -> build/lib.linux-x86_64-3.7/sasl running build_ext building 'sasl.saslwrapper' extension creating build/temp.linux-x86_64-3.7 creating build/temp.linux-x86_64-3.7/sasl gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Isasl -I/usr/local/include/python3.7m -c sasl/saslwrapper.cpp -o build/temp.linux-x86_64-3.7/sasl/saslwrapper.o unable to execute 'gcc': Permission denied error: command 'gcc' failed with exit status 1 [end of output]

![image](https://user-images.githubusercontent.com/67045657/229008581-63df06a5-0b0a-476b-96a6-2035a52174c6.png)

Other packages are fine after installation, example, docker, mssql, mongodb etc...

Can anyone help?

Thanks in advance

**Comment by potiuk** [comment, marked as answer]
Yes.  You need to aither follow the:

* extending image with packages requiring compilation following example here: https://airflow.apache.org/docs/docker-stack/build.html#example-when-you-add-packages-requiring-compilation (if you are not worried about image size)
* or build your own customized image from the scratch following examples here: https://airflow.apache.org/docs/docker-stack/build.html#examples-of-image-customizing (if you want to get highly optimized image as the result).

Context: the package you want to install does not have native binary package compiled for your version of python/architecture, so python wants to compile it - and since the base docker image has no compiller libraries and tools (which makes it 200 MB - some 25%  - smaller). If you need packages that need compilation, you can add those compiler libraries and tools but your image will be much bigger (you can read docker docs about optimizing file size to see what consequences it has, basically slower startup and bigger overhead when running).

On the other hand going "customization" rout allow you to follow what Airflow Docker image does internally - i.e. use multi-stage builds to get rid of the compiler tools and libraries in the target image.

More details and in-depth explanation (and comparisions of the two options) are here: https://airflow.apache.org/docs/docker-stack/build.html#extending-vs-customizing-the-image

**Reply by Kelvinchow79** [reply]
Thanks for the response, as i am still an entry level in devops, so my linux and docker are beginer level. I have not reach the level to edit or create and dockerfile yet. anyway, if i were to do that, i would first need to know which version of each package is needed. I have docker exec into each of the containers to pip install several providers, many are working, example databricks, docker etc, but i could not get to see snowflake to appear in UI after following all the requirements from the airflow documentation. if i were to follow your way to add into the docker compose image of airflow, what should i add for snowflake to work? thanks

**Reply by potiuk** [reply]
Because you are not supposed to install packages via execing to the containers. It won't work purely because this is not how docker works. your webserver interpreter will start when the container starts. You can take a look at our guides on "running airflow in docker"  https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html#using-custom-images and your ONLY reasonable option is to build custom images. There are some development tools to make iterations easier (for example if you look at the docs of ours docker-compose has the "build" feature that allow you to rebuild and use the image dynamically when you run it with `--build` command. 

just to be brutally honest I am afraid, you need to learn and master docker, docker images, building, registries and docker-compose itself if you want to use it. Those are tools that require to learn them if you use it, they are not working "out of the box", you need to deeply understand what you are doing, so make sure to have time on catching up. If you don't have time to do it - don't go docker route.

**Reply by Kelvinchow79** [reply]
I fully understand what you are saying, i appreciate it. thanks. I have one other question, after i built a custom image of airflow with those installation, the docker-compose will use that image, something like that?

**Reply by potiuk** [reply]
read  the docs -> they explain you how to make docker compose does.

**Reply by Kelvinchow79** [reply]
by the way, i have trouble finding information on the modification of the airflow config.py file. I know that it create a default one in each container upon first set up. how should i change the config file? i was thinking of mounting a file in docker host to the config file in each container. i place the modified file in docker host, then I docker-compose up again. can i do that? or is there a better way? thanks.

**Reply by potiuk** [reply]
>  have trouble finding information on the modification of the airflow config.py file

Because we doliberately do not want provide opinionated solution on how to do such things with such versatile developer tool like docker compose is.

I'd say again - it's up to you, really, Docker Compose is all yours to manage. We don't have a "ready-to-use" standard recipe that is the only way. Docker-compose is really a developer tool., that everone should tailor to their need. 

But yes - you could do that. You could also manage configuration via env vars (see conifg docs in airflow) - and again with docker-compose you can do it in multiple ways, via .env file, setting values in the env file, or setting empty var names in the .env file and letting docker-compose pass-through them to the container, or configuring them in your docker-compose file. You can even specify them in a different docker-compose file and add it to the command line so that the two (or more) compose files are merged, and you can even specify an env variable containing names of compose files to combine. etc. etc. 

It really depends on what use cases you are going to handle - and getting familiar with docker compose first is something I recommend.

**Reply by Kelvinchow79** [reply]
Thank you.

**Reply by Kelvinchow79** [reply]
Hi Potiuk,

I have read the documents and attempt to extend the image, and use it in the docker-compose, but init fail with exit 1, and only containers postgres and redis are up.
Dockerfile:
FROM apache/airflow:2.5.1
COPY requirements.txt /
RUN pip install --no-cache-dir -r /requirements.txt

requirements.txt:
awscli
sagemaker
snowflake-connector-python==2.4.1
apache-airflow-providers-snowflake==2.5.0
apache-airflow-providers-common-sql
apache-airflow-providers-docker
apache-airflow-providers-apache-cassandra
apache-airflow-providers-mongo
apache-airflow-providers-microsoft-mssql
apache-airflow-providers-apache-pig
apache-airflow-providers-presto
apache-airflow-providers-qubole
apache-airflow-providers-segment
apache-airflow-providers-apache-spark
apache-airflow-providers-vertica
snakebite-py3
apache-airflow-providers-databricks
argparse
hdfs
apache-airflow-providers-apache-hdfs==2.0.0

docker build . -f Dockerfile --tag airflow:dev (image build without error)

Line change in docker-compose:
image: ${AIRFLOW_IMAGE_NAME:-apache/airflow:2.5.1}
to:
image: airflow:dev

however it works after i change back to
image: apache/airflow:2.5.1 (which is the original image that is still residing in local)

May I know how do i troubleshoot this issue?

Thanks

**Reply by potiuk** [reply]
Generally - as with any other debugging - you need to look at the logs. You need to look at what error syou have when you start (error 1 does not show the error) and take a  look at the logs of your pip installation (so output of your docker build to see if anything unusal happens (after you see what is the error you see when you start airflow).

Just the usual things.
